---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.2
  kernelspec:
    display_name: Bash
    language: bash
    name: bash
---

```{bash}
source config.sh

```

# Check Strandness
`infer_experiment.py -r {genome_bed} -i {workdir}/{uid}/star/Aligned.sortedByCoord.out.bam`

# Picard CollectRnaSeqMetrics

`PicardCommandLine CollectRnaSeqMetrics I=/star_output/Aligned.sortedByCoord.out.bam O=/picard_output/RNA_Metrics.txt REF_FLAT=genome.refFlat STRAND=FIRST_READ_TRANSCRIPTION_STRAND RIBOSOMAL_INTERVALS=genome.intervalList`




 Yes we do have this QC pipeline, and applied it to a recent project,

https://gitlab.oit.duke.edu/dcibioinformatics-internal/projects/freedman-allen-prostate/-/tree/master/

Basically you can find all the commands we actually run in the 'recipe' files, software configurations and paths in 'settings' files, and sample-level settings in 'samples' files. Here're the steps:

 1.  Generate STAR index and parse GTF files using `dcibioinformatics-star-reference/ref.*`
 2.  Run STAR/Picard using `dcibioinformatics-rnaseq/stage2.*` (stage1.* were for fastqc and trimming steps)

 To generate a report using STAR/Picard outputs, you may want to have a look at this R tool created by Jeff. It can be installed from this repo:

https://gitlab.oit.duke.edu/dcibioinformatics-internal/soft/dcibioinformatics

and the tutorial can be found at the wiki page,

https://gitlab.oit.duke.edu/dcibioinformatics-internal/soft/dcibioinformatics/-/wikis/Vignettes

