# Duke Microbiome Immunology Cancer Bioinformatics Short Course


In this course, you will learn how to generate and analyze RNAseq data. Roughly, we have the following components:

- *Experimental Design and Statistics* How do we design experiments so that results are easily interpretable and answer the question(s) we are interested in. 

- *Analysis of Data* How do we **properly** analyze experimental data, so that results are correct.

- *Computational Procedure* Analysis pipeline (Bioinformatics)

Finally, we want to do all of the above in a **REPRODUCIBLE** fashion. 


## Computational bootcamp

This 'computational bootcamp' component is designed to give you the tools you need to carry out the steps above - and to do so within the context of reproducible research.

The analysis pipeline has several different apps. Some are written in R, and so require knowledge of the R programming language (minimal - but some proficiency is needed). Some have components in python, and some are binaries. Both of these last types of applications require moderate proficiency in the 'bash shell' or 'unix command line'. Therefore, we will cover the following topics:

- Basic R
- Basic Unix/Linux commands

Additionally, we will use the bootcamp to reinforce the statistical lecture materials by walking you through some of the examples using R, and we will cover some 'data visualization' techniques that include graphics in R.

All of this will be done within the Jupyter notebook tool, which allows for what is called 'literate programming' and reproducible pipelines.

We execute code cells (and markdown cells) by hitting &lt;shift>&lt;Enter>. Take a moment to play around a bit and get comfortable. Click on the 'help' menu and look at the keyboard shortcuts (they are very useful).

```{r}

```


```{r}
# Make the next cell into a markdown cell (using the keyboard shortcut) and type some text. 
# Can you make it bold? Italicized? 
# Can you figure out how to make headings?
```


```{r}

```

## Code cells

```{r}
# This is a code cell. The notebook is using an R kernel to interpret whatever is entered here. (See the upper right corner)

2+2
```

Note that we can change kernels (kernel tab above). Installed on this container are R, Bash and Python 3.

```{r}
# I can execute R code (just like the 2+2 above)
hist(rnorm(100))
```

```{r}
# We will use dplyr, tidyr and ggplot extensively
library(tidyverse)
```

```{r}
# R has built-in data sets we can play with
head(mtcars)
```

```{r}
mtcars %>% ggplot(aes(x = hp, y = mpg, color = as_factor(cyl))) + geom_point()
```

```{r}
mtcars %>% ggplot(aes(x = hp, y = mpg, color = as_factor(cyl))) + geom_point() + geom_smooth(method = 'lm')
```

```{r}

```
