---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.2
  kernelspec:
    display_name: R
    language: R
    name: ir
---

# Import Count files


The objective of this notebook is to import the project's STAR count files into an R object.


### Load packages

```{r}
library(tidyverse)
library(foreach)
```

### Set directory containing count files

```{r}
cntdir <- normalizePath("/data/starout/")

# Output directory for R image files
outdir <- "./proc/"

if (!dir.exists(outdir)) {
  dir.create(outdir)
}

```

### Get count file names

```{r}
cntfiles <- list.files(path = cntdir, pattern = "*_ReadsPerGene.out.tab")
```

### Inspect file names

```{r}
cntfiles
```

## Import a single count file


### Get full path name for first two count files

```{r}
testfile1 <- file.path(cntdir, cntfiles[1])
testfile1
testfile2 <- file.path(cntdir, cntfiles[2])
testfile2
```

### Import counts for test files

```{r}
readr::read_tsv(testfile1, col_names = FALSE) -> testcnt1
readr::read_tsv(testfile2, col_names = FALSE) -> testcnt2
```

### Inspect first ten rows of first count file


First look at the dimension and note that there are four columns

```{r}
testcnt1 %>% dim()
```

Next, inspect the first ten rows. Columns 2 through 4 contain read counts. The choice for downstream analysis depends on the stranding of kit (we will use the second column as we are analyzing data from an unstranded kit). The first four rows, provide the corresponding counts for unmapped, multi-mapped, no-feature-mapped and ambigiously-mapped counts. The remaining rows, provide the gene counts.

From the STAR manual

Counting number of reads per gene. With --quantMode GeneCounts option STAR will count number reads per gene while mapping. A read is counted if it overlaps (1nt or more) one and only one gene. Both ends of the pairedend read are checked for overlaps. The counts coincide with those produced by htseq-count with default parameters. This option requires annotations (GTF or GFF with â€“sjdbGTFfile option) used at the genome generation step, or at the mapping step. STAR outputs read counts per gene into ReadsPerGene.out.tab file with 4 columns which correspond to different strandedness options:

*column 1: gene ID

*column 2: counts for unstranded RNA-seq

*column 3: counts for the 1st read strand aligned with RNA (htseq-count option -s yes)

*column 4: counts for the 2nd read strand aligned with RNA (htseq-count option -s reverse)

```{r}
testcnt1 %>% head(10)
```

The counts across the rows of each of the three columns (as expected) are identical 

```{r}
testcnt1 %>% dplyr::summarize(across(2:4,sum))
```

### Inspect first ten rows of second count file


Note that the structure of the file is identical to that of the first count files (only the counts, as expected, differ).

```{r}
testcnt2 %>% head(10)
```

### Extract first and fourth columns of first count file

```{r}
testcnt1 %>% 
    dplyr::select(X1, X2) %>% 
    head(10)
```

### Extract and rename first and fourth columns of first count file

```{r}
testcnt1 %>% 
    dplyr::select(X1, X2) %>% 
    rlang::set_names(c("gene", basename(testfile1))) %>%
    head(10)
```

### Extract and rename first and fourth columns of first and second count files and merge by gene name

```{r}
testcnt1 %>% 
    dplyr::select(X1, X2) %>% 
    rlang::set_names(c("gene", basename(testfile1))) ->
    cnt1

testcnt2 %>% 
    dplyr::select(X1, X2) %>% 
    rlang::set_names(c("gene", basename(testfile2))) ->
    cnt2

dplyr::full_join(cnt1, cnt2, by = "gene") %>% head(10)
```

### Silence dplyr::read_csv by explicitely defining the column types

```{r}

### Data type for each column in STAR count file (gene name is character; the three remaining columns are integers)
coltypes <- list(col_character(), col_integer(), col_integer(), col_integer())
readr::read_tsv(testfile1, col_names = FALSE, col_types = coltypes) %>% head(10)
```

## Import count files

```{r}
### Iteratively merge count files
countdata <- foreach(fname = cntfiles, .combine = function(df1,df2){dplyr::full_join(df1, df2, by = "gene")}) %do% {
    cntfile <- file.path(cntdir, fname)
    readr::read_tsv(cntfile, col_names = FALSE, col_types = coltypes) %>%
    dplyr::select(X1, X2) %>%
    rlang::set_names(c("gene", fname))
}
```

The 12 count will be merged by gene name. Without the .combine statement, the loop will read in a list of 12 files. With the .combine statement the files will be merge iteratively as follows

* Merge {f1} and {f2} to get {f1,f2}
* Merge {f1,f2} and {f3} to get {f1,f2,f3}
* Merge {f1,f2,f3} and {f4} to get {f1,f2,f3,f4}
* ....
* Merge {f1,f2,f3,...,f11} and {f12} to get {f1,f2,f3,...f11,f12}


Inspect the dimension of the object

```{r}
countdata %>% dim
```

Inspect first 10 rows of the object

```{r}
countdata %>% head(10)
```

```{r}
countfile <- file.path(outdir, "PRJNA668393-countdata.RData")

saveRDS(countdata, file = countfile)

tools::md5sum(countfile)
```

```{r}
sessionInfo()
```
