---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.2
  kernelspec:
    display_name: R
    language: R
    name: ir
---

# Differential Expression Analysis with DESeq2


Carry out a basic set of DE analyses using DESeq2 and visualize the results

Requires the following:
- ./proc//PRJNA668393-ddsdata.RDS


## Load packages and data object


### Load packages

```{r}
library(tidyverse)
library(DESeq2)
```

### Load the PRJNA668393 dds object from image file

```{r}
procdir <- normalizePath("./proc/")
ddsfile <- file.path(procdir, "PRJNA668393-ddsdata.RDS")

PRJNA668393ddsData <- readRDS(ddsfile)

tools::md5sum(ddsfile)
```


# Inspect object & Slots of an S4 class


Let's have a look at the object we have created.

```{r}
PRJNA668393ddsData
```

see the class of dds object

```{r}
class(PRJNA668393ddsData)
```

DESeqDataSet is an S4 object. Recall that an S4 object was taught when introducing bioconductor. Note that S4 objects allow users to wrap up multiple elements into a single variables where each element is called a slot.

```{r}
slotNames(PRJNA668393ddsData)
```

The metadata (columnData) is stored in the slot `colData`

```{r}
SummarizedExperiment::colData(PRJNA668393ddsData) %>% as.data.frame
```

The design formula is stored in the slot `design`. The design holds the R formula which expresses how the counts depend on the variables in colData. We have specified any predictor values, so this currently looks like an intercept-only model.

```{r}
DESeq2::design(PRJNA668393ddsData)
```

The first thing you may want to do is **have a look at the raw counts** you have imported. The `DESeq2::counts` function extracts a matrix of counts (with the genes along the rows and samples along the columns). Let us first verify the dimension of this matrix.

```{r}
dim(DESeq2::counts(PRJNA668393ddsData))
```

```{r}
head(DESeq2::counts(PRJNA668393ddsData),3)
```

This slot is empty, but it will be important soon. This is where dispersion estimates for variation in read counts per gene will go, i.e. dispersion parameter $\alpha_i$ (notation below).

```{r}
dispersionFunction(PRJNA668393ddsData)
```

This slot is also empty, but it too will be important soon. This is where sample-specific size factors will go, ie ùë†ùëó  is the sample ùëó specific size factor (notation below). A size factor is a normalization constant used to account for the fact that samples can have very different library sizes
```{r}
sizeFactors(PRJNA668393ddsData)
```
# Differential expression analysis


## Specify Model


**Summary of notation**
- $K_{ij}$ denotes the observed **number of reads** mapped to gene $i$ for sample $j$
- $K_{ij}$ follows a **negative binomial distribution** with
    - **Mean** $\mu_{ij}$
    - **Dispersion parameter** $\alpha_i$
    
- Model:
  - $K_{ij} \sim NB(\mu_{ij}, \alpha_i)$ (<em>i.e.,</em> the mean number of reads from library $j$ mapped to gene $i$ follows a negative binomial distribution with mean $\mu_{ij}$ and dispersion parameter $\alpha_i$)
  - $\mu_{ij} = s_{j}q_{ij}$
  - $s_j$ is sample $j$ specific normalization constant to account for the fact that samples can have very different library sizes
  - $q_{ij}$ is  gene $i$ and sample $j$ specific estimated mean reads after accounting for sample-specific library size
  - $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} x_j + \beta_{2i} z_j $
       - $x_j=0$ if treament = isotype, or
       - $x_j=1$ if treatment = anti-PD1
       - $z_j=0$ if cell line = Parental, or
       - $z_j=1$ if cell line = UV2.
  
- **Linear predictor**
 -  $\log_2(q_{ij}|(x_j=0, z_j=0)) = \beta_{0i}$
 -  $\log_2(q_{ij}|(x_j=1, z_j=0)) = \beta_{0i}  + \beta_{1i}$
 -  $\log_2(q_{ij}|(x_j=0, z_j=1)) = \beta_{0i}  + \beta_{2i}$
 -  $\log_2(q_{ij}|(x_j=1, z_j=1)) = \beta_{0i}  + \beta_{1i} +  + \beta_{2i}$
 
- **Mean count** 
 -  $q_{ij}|(x_j=0, z_j=0) = 2^{\beta_{0i}}$
 -  $q_{ij}|(x_j=1, z_j=0) = 2^{\beta_{0i}  + \beta_{1i}}$
 -  $q_{ij}|(x_j=0, z_j=1) = 2^{\beta_{0i}  + \beta_{2i}}$
 -  $q_{ij}|(x_j=1, z_j=1) = 2^{\beta_{0i}  + \beta_{1i} +  + \beta_{2i}}$
     
- **Normalized mean count**
 -  $\mu_{ij} | (x_j=0,z_j=0) = s_{j} * 2^{\beta_{0i}}$
 -  $\mu_{ij} | (x_j=1,z_j=0) = s_{j} * 2^{\beta_{0i} + \beta_{1i}}$
 -  $\mu_{ij} | (x_j=0,z_j=1) = s_{j} * 2^{\beta_{0i} + \beta_{2i}}$
 -  $\mu_{ij} | (x_j=1,z_j=1) = s_{j} * 2^{\beta_{0i} + \beta_{1i} + \beta_{2i}}$
  
- **Effect size**
  - Treatment effect within the parental cell line ($z_j=0$)
	- $\Delta_{i0} = (\beta_{0i} + \beta_{1i}) - (\beta_{0i}) = \beta_{1i}$
  - Treatment effect within the UV2 cell line ($z_j=1$)
	- $\Delta_{i1} = (\beta_{0i} + \beta_{1i} + \beta_{2i}) - (\beta_{0i} + \beta_{2i}) = \beta_{1i}$

- **Parameter interpretation**
	- $\beta_{1i}$ quantifies the treatment effect
	- Under the additive model, the treatment effect does not depend on the cell line: $\Delta_{0i} = \Delta_{1i} = \beta_{i1}$
	


## Set up data object for analysis


### Specify model design


Add the design to the DESeq2 object. This is going to be the format that R uses for most linear models with the form "response ~ fixedeffect1 + fixedeffect2".

```{r}
ddsobj <- PRJNA668393ddsData
```

```{r}
design(ddsobj) <- formula(~ cline + trt)
```

```{r}
design(ddsobj)
```

### Size Factors
 We begin by estimating the size factors for each sample $s_1,\ldots,s_n$:

```{r}
ddsobj <- DESeq2::estimateSizeFactors(ddsobj)
```

Now, compare the dds object to that of before applying the estimateSizeFactors() function. What has changed? What remains unchanged?

```{r}
ddsobj
```

Note that there is a **sizeFactor** added to **colData**. Let's look at it more carefully

You can also get the size factors directly

```{r}
tibble::tibble(sfactor = sizeFactors(ddsobj))
```

Summarize size factors

```{r}
summary(DESeq2::sizeFactors(ddsobj))
```

Now that the size factors have been estimated, we can get "normalized" read counts. Remember that DESeq2 normalizes against size factor.

```{r}
# original counts for three genes for samples 3 and 11
DESeq2::counts(ddsobj)[102:104,c(3,11)]
```

```{r}
# Size factors for samples 3 and 11

DESeq2::sizeFactors(ddsobj)[c(3,11)]
```

```{r}
# normalized count for samples 3 and 11
DESeq2::counts(ddsobj, normalize = TRUE)[102:104, c(3,11)]
```

```{r}
# normalized manually using size factors for sample 3
DESeq2::counts(ddsobj)[102:104, 3] / DESeq2::sizeFactors(ddsobj)[3]
```

```{r}
# normalized manually using size factors for sample 11
DESeq2::counts(ddsobj)[102:104, 11] / DESeq2::sizeFactors(ddsobj)[11]
```

Here is how you can extract the raw and normalized counts for specific genes and samples

```{r}
counts(ddsobj, normalize = FALSE)[c("ENSMUSG00000061024.9", "ENSMUSG00000079671.9"),c("SRR12804465","SRR12804467")]
```

```{r}
counts(ddsobj, normalize = TRUE)[c("ENSMUSG00000061024.9", "ENSMUSG00000079671.9"),c("SRR12804465","SRR12804467")]
```

### Dispersion Parameters
Next, we get the dispersion factors for each gene $\alpha_1,\ldots,\alpha_{m}$

```{r}
ddsobj <- estimateDispersions(ddsobj)
```

Now inspect the dds object again and note that the rowRanges slot has extra information ("metadata column names(0):" before versus "column names(9): baseMean baseVar ... dispOutlier dispMAP")
- before: 
    - `metadata column names(0):`
- after:  
    - `column names(9): baseMean baseVar ...`

```{r}
ddsobj
```

Can you notice the difference?



Note that the dispersionFunction slot is now populated.

```{r}
dispersionFunction(ddsobj)
```

The dispersion function slot holds the fitted model used to estimate gene-specific dispersion. For the negative binomial distribution, the variance is equal to 

$\sigma^2 = \mu_{ij} + \mu_{ij}^2 /\alpha_i$

where $\alpha_i$ is the dispersion parameter. **I'd like more info about how this is represented in the dispersionFunction slot** 

We can extract the gene specific dispersion factors using dispersions(). 

```{r}
alphas <- dispersions(ddsobj)
alphas[1:10]
```

**Why are some of these dispersion parameter estimates "NA"?**


Verify that the number of dispersion factors equals the number of genes

```{r}
# number of disperion factors
length(alphas)
```

```{r}
summary(alphas)
```

Summarize the dispersion factors using a box plot (may want to log transform)

```{r}
boxplot(log(dispersions(ddsobj)))
```

# Differential Expression Analysis
We can now conduct a differential expression analysis using the DESeq() function. Keep in mind that to get to this step, we first estimated the size factors for each sample and then the dispersion parameters for each gene.

```{r}
### Carry out DE analysis
ddsDE <- DESeq2::DESeq(ddsobj)
```

```{r}
### Look at object
ddsDE
```

```{r}
### Look at some of the results
DESeq2::results(ddsDE)
```

### Look at some of the results (tidy version sorted by adjusted P-value)

```{r}
DESeq2::results(ddsDE, tidy = TRUE) %>% 
    dplyr::arrange(padj) %>%
        head(10)
```

We can get the results for the differential expression analysis using results(). Be careful because how we define our contrasts has a huge impact on how we interpret the results. 

Note that if you do not define the contrast, the default constrast is the last term in your additive model `design(dds)`. For us, this is the cell line (cline) but that is probably not the first contrast you are interested in examining. 

We will take a look at the treatment effect first by specifying "trt" as a the first element in the vector passed to the contrast argument within results(). Be aware that the order you specify the treatment levels really matters a lot for interpretation of the results.

Here, we specify that "isotype" is the reference level (i.e. $x_j=0$ if treament = isotype) so positive effect sizes indicate that treatment with anti-PD-1 up-regulates the gene of interest.
```{r}
# DE with respect to treatment using isotype as the referent
# (positive effect size suggests that treatment up-regulates)
DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"))
```

Alternatively, we could specify that "antiPD1" is the reference level (i.e. $x_j=0$ if treament = antiPD1) so negative effect sizes indicate that treatment with antiPD1 up-regulates the gene of interest.
```{r}
# DE with respect to treatment using anti-PD-1 as the referent 
# (negative effect size suggests that treatment up-regulates)
DESeq2::results(ddsDE, contrast = c("trt", "isotype", "antiPD1"))
```

Let's look at the results for the first four genes arranged by p-value for each contrast in the model:

Effect of treatment (trt) using "isotype" as the reference
```{r}
### Tidy the results
DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>% 
    dplyr::arrange(padj) %>%
        head(10)
```

Effect of cell line (cline) using "Parental" as the reference
```{r}
### DE effect with respect to cell line
### Tidy the results
DESeq2::results(ddsDE, contrast = c("cline", "UV2", "Parental"), tidy = TRUE) %>% 
    dplyr::arrange(padj) %>%
        head(10)
```

### Extract and interpret coefficient estimates


The results for the top hit (gene with the lowest p-value) for treatment DE are shown below

```{r}
DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>%
    dplyr::filter(row == "ENSMUSG00000110537.2")
```

This output only provides the estimate for the treament effect $\beta_{1i}$ for this gene. We can get the estimates for the intercept $\beta_{0i}$ and the cell line effect $\beta_{2i}$ as follows

```{r}
coef(ddsDE, se = TRUE)["ENSMUSG00000110537.2", , drop = FALSE]
```

- Parameter estimates for ENSMUSG00000110537.2

    * $\hat{\beta}_{i0} = 5.240018$
    * $\hat{\beta}_{i1} = -1.302921$
    * $\hat{\beta}_{i2} = -0.2218365$


- Estimated mean counts

    * trt == 0 and cline == 0  => $\log2(q_{ij})$= 5.240018	+ (0) + (0)  = 5.240018
    * trt == 1 and cline == 0  => $\log2(q_{ij})$= 5.240018	+ (-1.302921) + (0) = 3.937097
    * trt == 0 and cline == 1  => $\log2(q_{ij})$= 5.240018	+ (0) + (-0.2218365) = 5.018181
    * trt == 1 and cline == 1  => $\log2(q_{ij})$= 5.240018	+ (-1.302921) + (-0.2218365) = 3.715261

- The treatment DE effect under the additive model does not depend on cell line

    * if cline == 0 => $\hat{\Delta}_{i0}$ = 3.937097 - 5.240018  = -1.302921
    * if cline == 1 => $\hat{\Delta}_{i1}$ = 3.715261 - 5.018181 = -1.30292

**As exepected $\hat{\Delta}_{i0} = \hat{\Delta}_{i0} = \hat{\beta}_{i1}$**



## Visualize DE effect


### Visualize top two hits


Looking at the results for these two genes: 

* The estimated log2FC for "ENSMUSG00000110537.2" is negative. We will verify visually that treatment is associated with lower expression

* The estimated log2FC for "ENSMUSG00000022564.8" is positive. We will verify visually that treatment is associated with higher expression



```{r}
DESeq2::plotCounts(ddsobj, "ENSMUSG00000110537.2", intgroup = "trt", normalized = TRUE)

DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>%
    dplyr::filter(row == "ENSMUSG00000110537.2")
```

```{r}
DESeq2::plotCounts(ddsobj, "ENSMUSG00000022564.8", intgroup = "trt", normalized = TRUE)

DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>%
    dplyr::filter(row == "ENSMUSG00000022564.8")
```

DESeq2::plotCounts() is a convenient function for visualizing differential expression effects. The following example illustrates how one could customize the figure

```{r}
myinteractplot <- function(mydds, geneid) {
    
    SummarizedExperiment::assay(mydds) %>%
        tibble::as_tibble(rownames="gene") %>%
            dplyr::filter(gene == geneid) %>%
                tidyr::gather(Run, value, -gene) %>%
                    dplyr::select(-gene) -> 
                        expdat
    
    SummarizedExperiment::colData(mydds) %>%
        as.data.frame %>%
            tibble::as_tibble() %>%
                dplyr::full_join(expdat, by = "Run") -> 
                    genedat
    
    genedat %>%
        ggplot2::ggplot(aes(x = trt, y = value/sizeFactor, color = cline)) +
            ggplot2::geom_point() + 
                ggplot2::xlab("Treatment") + ggplot2::ylab(paste(geneid, "(normalized count)")) +
                ggplot2::scale_colour_manual(name = "", values = c("red3", "blue3")) +
                    ggplot2::theme_bw()
    }

myinteractplot(ddsobj, "ENSMUSG00000110537.2")

myinteractplot(ddsobj, "ENSMUSG00000022564.8")
```


Bonus exercise: The previous helper function allowed us to customize the differential expression visualization. It is however in need of improvement and refinement if to be used for other projects. Specifically, there is substantial hard coding (e.g., the variable name for the merger "Run", the experimental factors "trt" and "cline", the axis labels, the colors. Consider revising this function so that it can be used more generally.


### Volcano plot


Volcano plots can be useful as a visual summary of how many genes respond to a treatment because we can simultaneously see how gene responses vary by effect size (x-axis) and significance (y-axis).

Can you think of reasons why we might see some genes that have large effect sizes but higher p-values? How about small effects sizes and lower p-values?

```{r}
### Volcano plot for treatment effect
DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>%
    dplyr::filter(!is.na(pvalue)) %>%
        ggplot2::ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) + 
            ggplot2::geom_point() + 
                ggplot2::theme_bw()
```

```{r}
### Volcano plot for cell line effect
DESeq2::results(ddsDE, contrast = c("cline", "UV2", "Parental"), tidy = TRUE) %>%
    dplyr::filter(!is.na(pvalue)) %>%
        ggplot2::ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) + 
            ggplot2::geom_point() + 
                ggplot2::theme_bw()
```

## Save results

```{r}
outdir <- "./proc"
outfile <- file.path(outdir, "PRJNA668393-DESeq2-additive.RDS")

saveRDS(ddsDE, file = outfile)

tools::md5sum(outfile)
```

```{r}
sessionInfo()
```
