---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.2
  kernelspec:
    display_name: R
    language: R
    name: ir
---

# Interaction Analysis with DESeq2


Carry out a basic set of interaction analyses using DESeq2 and visualize the results

Requires the following:
- ./proc//PRJNA668393-ddsdata.RDS



## Load packages and data objects


### Load packages

```{r}
library(tidyverse)
library(DESeq2)
```

### Load the PRJNA668393 dds object from image file

```{r}
procdir <- normalizePath("./proc/")
ddsfile <- file.path(procdir, "PRJNA668393-ddsdata.RDS")

PRJNA668393ddsData <- readRDS(ddsfile)

tools::md5sum(ddsfile)
```

## Specify Model


**Summary of notation**
- $K_{ij}$ denotes the observed **number of reads** mapped to gene $i$ for sample $j$
- $K_{ij}$ follows a **negative binomial distribution** with
    - **Mean** $\mu_{ij}$
    - **Dispersion parameter** $\alpha_i$
    
- **Model**:
  - $K_{ij} \sim NB(\mu_{ij}, \alpha_i)$ (<em>i.e.,</em> the mean number of reads from library $j$ mapped to gene $i$ follows a negative binomial distribution with mean $\mu_{ij}$ and dispersion parameter $\alpha_i$)
  - $\mu_{ij} = s_{j}q_{ij}$
  - $s_j$ is sample $j$ specific normalization constant to account for the fact that samples can have very different library sizes
  - $q_{ij}$ is  gene $i$ and sample $j$ specific estimated mean reads after accounting for sample-specific library size
  - $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} x_j + \beta_{2i} z_j + \beta_{3i} (x_j*z_j)$
       - $x_j=0$ if treament = isotype, or
       - $x_j=1$ if treatment = anti-PD1
       - $z_j=0$ if cell line = Parental, or
       - $z_j=1$ if cell line = UV2.
  
- **Linear predictor**
 -  $\log_2(q_{ij}|(x_j=0, z_j=0)) = \beta_{0i}$
 -  $\log_2(q_{ij}|(x_j=1, z_j=0)) = \beta_{0i}  + \beta_{1i}$
 -  $\log_2(q_{ij}|(x_j=0, z_j=1)) = \beta_{0i}  + \beta_{2i}$
 -  $\log_2(q_{ij}|(x_j=1, z_j=1)) = \beta_{0i}  + \beta_{1i} +   \beta_{2i} + \beta_{3i} $
 
- **Mean count** 
 -  $q_{ij}|(x_j=0, z_j=0) = 2^{\beta_{0i}}$
 -  $q_{ij}|(x_j=1, z_j=0) = 2^{\beta_{0i}  + \beta_{1i}}$
 -  $q_{ij}|(x_j=0, z_j=1) = 2^{\beta_{0i}  + \beta_{2i}}$
 -  $q_{ij}|(x_j=1, z_j=1) = 2^{\beta_{0i}  + \beta_{1i} +  \beta_{2i}+ \beta_{3i} }$
     
- **Normalized mean count**
 -  $\mu_{ij} | (x_j=0,z_j=0) = s_{j} * 2^{\beta_{0i}}$
 -  $\mu_{ij} | (x_j=1,z_j=0) = s_{j} * 2^{\beta_{0i} + \beta_{1i}}$
 -  $\mu_{ij} | (x_j=0,z_j=1) = s_{j} * 2^{\beta_{0i} + \beta_{2i}}$
 -  $\mu_{ij} | (x_j=1,z_j=1) = s_{j} * 2^{\beta_{0i} + \beta_{1i} + \beta_{2i}+ \beta_{3i} }$
  
- **Effect size**
  - Treatment effect within the parental cell line ($z_j=0$)
	- $\Delta_{i0} = (\beta_{0i} + \beta_{1i}) - (\beta_{0i}) = \beta_{1i}$
  - Treatment effect within the UV2 cell line ($z_j=1$)
	- $\Delta_{i1} = (\beta_{0i} + \beta_{1i} + \beta_{2i}+ \beta_{3i}) - (\beta_{0i} + \beta_{2i}) = \beta_{1i} + \beta_{3i}$

- **Parameter interpretation**
	- $\beta_{1i}$ quantifies the treatment effect for the Parental cell line
    - $\beta_{1i} + \beta_{3i}$ quantifies the treatment effect for the UV2 cell line
	- Under the multiplicative model, the treatment effect depends on the cell line: $\beta_{i1}=\Delta_{0i} \ne \Delta_{1i} = \beta_{i1}+ \beta_{3i}$
    - The multiplicative model reduces to an additive model if $\beta_{3i} = 0$


## Set up data object for analysis


### Specify model design

```{r}
ddsobj <- PRJNA668393ddsData
```

```{r}
design(ddsobj) <- formula(~ cline + trt + cline:trt)
```

```{r}
design(ddsobj)
```

### Size Factors
 We begin by estimating the size factors $s_1,\ldots,s_n$:

```{r}
ddsobj <- estimateSizeFactors(ddsobj)
```

### Dispersion Parameters
Next, we get the dispersion factors $\alpha_1,\ldots,\alpha_{m}$

```{r}
ddsobj <- estimateDispersions(ddsobj)
```

## Interaction Analysis

```{r}
### Carry out interaction analysis
ddsDE <- DESeq2::DESeq(ddsobj)
```

```{r}
### Look at object
ddsDE
```

```{r}
## Model parameters 
DESeq2::resultsNames(ddsDE)
```

Remember these correspond to
- Intercept = $\beta_{0i}$
- cline_UV2_vs_Parental = $\beta_{1i}$
- trt_antiPD1_vs_isotype = $\beta_{2i}$
- clineUV2.trtantiPD1 = $\beta_{3i}$

```{r}
### Look at some of the results
DESeq2::results(ddsDE)
```

### Look at some of the results (tidy version sorted by adjusted P-value)

```{r}
DESeq2::results(ddsDE, tidy = TRUE) %>% 
    dplyr::arrange(padj) %>%
        head(10)
```

We can get the results for the differential expression analysis using results(). Here, we can compare two group of samples specified by the contrast. (If not, the default contrast would be the last term in your additive model `design(dds)`).


### Extract coefficient estimates


The results for two genes are shown below

```{r}
DESeq2::results(ddsDE, tidy = TRUE) %>%
    dplyr::filter(row %in% c("ENSMUSG00000022351.15","ENSMUSG00000025964.16"))
```

This output only provides the estimate for the treament effect $\beta_{1i}$ for this gene. We can get the estimates for the intercept $\beta_{0i}$ and the cell line effect $\beta_{2i}$ as follows

```{r}
coef(ddsDE)[c("ENSMUSG00000022351.15","ENSMUSG00000025964.16"), ]
```

- Parameter estimates for ENSMUSG00000022351.15

* $\hat{\beta}_{i0} = 10.511809$
* $\hat{\beta}_{i1} = -0.1326850$
* $\hat{\beta}_{i2} = -0.4110139$
* $\hat{\beta}_{i3} = -1.350313$

- Estimated mean counts

* trt == 0 and cline == 0  => $\log2(q_{ij})$= 10.511809	
* trt == 1 and cline == 0  => $\log2(q_{ij})$= 10.511809 + (-0.1326850) = 10.37912
* trt == 0 and cline == 1  => $\log2(q_{ij})$= 10.511809 + (-0.4110139) = 10.1008
* trt == 1 and cline == 1  => $\log2(q_{ij})$= 10.511809 + (-0.1326850) + (-0.4110139) +(-1.350313) = 8.617797

The treatment DE effect under the multiplicative model depends on cell line: Under cell line 0, the treatment effect is the cell line log2 fold change, under cell line 1, the treatment effect is the cell line loig2 fold chnage + the interaction log 2 fold change

* if cline == 0 => $\hat{\Delta}_{i0}$ = 10.37912 - 10.511809 = -0.132689
* if cline == 1 => $\hat{\Delta}_{i1}$ = 8.617797 - 10.1008 = -1.483003

Note that -1.483003 is equal to $\beta_{2i}$(trt_antiPD1_vs_isotype = -0.132689) plus $\beta_{3i}$(clineUV2.trtantiPD1 = -1.350313)



We will now show how to get the within cell line treatment effects directly (this approach is recommended by the DESeq2 developer)


First add group variable combining the cell line and treatment effect

```{r}
dds1 <- ddsobj
dds1$group <- as.factor(paste0(dds1$trt,dds1$cline))
```

A look at the new group variable

```{r}
SummarizedExperiment::colData(dds1) %>% data.frame %>% dplyr::select(Run, trt, cline, group)
```

Conduct differential expression analysis using group

```{r}
DESeq2::design(dds1) <- ~ group
dds1mod <- DESeq2::DESeq(dds1)
```

```{r}
results(dds1mod)
```

```{r}
DESeq2::resultsNames(dds1mod)
# instead of 2 factors with two levels each, we now have 1 factor with 4 levels
```

Get the treament effect within the parental cell line

```{r}
DESeq2::results(dds1mod, tidy = TRUE, contrast = c("group","antiPD1Parental","isotypeParental")) %>%
    dplyr::filter(row %in% c("ENSMUSG00000022351.15"))
```

Get the treament effect within the UV2 cell line

```{r}
DESeq2::results(dds1mod, tidy = TRUE, contrast = c("group","antiPD1UV2","isotypeUV2")) %>%
    dplyr::filter(row %in% c("ENSMUSG00000022351.15"))
```

## Visualize DE effect


### Visualize top two hits

```{r}
myinteractplot <- function(mydds, geneid) {
    
    SummarizedExperiment::assay(mydds) %>%
        tibble::as_tibble(rownames="gene") %>%
            dplyr::filter(gene == geneid) %>%
                tidyr::gather(Run, value, -gene) %>%
                    dplyr::select(-gene) -> 
                        expdat
    
    SummarizedExperiment::colData(mydds) %>%
        as.data.frame %>%
            tibble::as_tibble() %>%
                dplyr::full_join(expdat, by = "Run") -> 
                    genedat
    
    genedat %>%
        ggplot2::ggplot(aes(x = cline, y = value/sizeFactor, color = trt)) +
            ggplot2::geom_point() + 
                ggplot2::xlab("Cell line") + ggplot2::ylab(paste(geneid, "(normalized count)")) +
                ggplot2::scale_colour_manual(name = "", values = c("red3", "blue3")) +
                    ggplot2::theme_bw()
    }

myinteractplot(ddsobj, "ENSMUSG00000022351.15")

myinteractplot(ddsobj, "ENSMUSG00000025964.16")
```


### Volcano plot

```{r}
### Volcano plot for treatment effect
DESeq2::results(ddsDE, tidy = TRUE) %>%
    dplyr::filter(!is.na(pvalue)) %>%
        ggplot2::ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) + 
            ggplot2::geom_point() + 
                ggplot2::theme_bw()
```

## Save results

```{r}
outdir <- "./proc"
outfile <- file.path(outdir, "PRJNA668393-DESeq2-multiplicative.RDS")

saveRDS(ddsDE, file = outfile)

tools::md5sum(outfile)
```

```{r}
sessionInfo()
```
