# The Unix Shell: Working with Text

## Text streams

Input and output of most Unix shell programs consists of plain text streams. Text output from a program can be piped into another program, or redirected to other streams. The standard streams are `stdin (0)` (standard input), `stdout (1)` (standard output) and `stderr (2)` (standard error). The default is to assume that input comes from `stdin` and output goes to `stdout`. We can also stream to and from a file. 

## Pipes and redirection

### Creating a text file from command line

For simple file creation, we can just use re-direction

A single `>` will create a new file or over-write an existing one.


```{bash}
echo "1 Hello, {bash}" > hello.txt
```


```{bash}
ls *txt
```

### Appending


```{bash}
echo "2 Hello, again" >> hello.txt
```

#### Special non-printing characters


```{bash}
echo -e "3 Hello\n4 again" >> hello.txt
```

### From file to `stdout


```{bash}
cat hello.txt
```

### Pipe to `cut` program to extract columns 2,3,4,5


```{bash}
cat hello.txt | cut -c 2-5
```

     Hel
     Hel
     Hel
     aga


### Building a chain of pipes

`wc -lc` reports the number of lines and bytes (usually corresponds to characters when using English text). 

Note that character count is 5 per line and not 4 because cut adds a newline character for each line.

#### Extract columns 2-5 and then count the number of lines and characters 


```{bash}
cat hello.txt | cut -c 2-5 | wc -lc  
```

          4      20


#### Extract lines 2-3


```{bash}
cat hello.txt | head -n 3 | tail -n 2
```

### Capturing error messages

The redirection operator > is actually 1> - that is, using stdout. We can also use 2> to redirect the output of stderr. &> means redirect both stdout and stderr, and is useful if for example, you want to direct all output to the same log file for later inspection.


```{bash}
mkdir foo/bar/baz > 'stdout.txt'
```

#### As there is nothing from stdout the file is empty


```{bash}
cat 'stdout.txt'
```

#### We need to use 2> to capture the output from stderr


```{bash}
mkdir foo/bar/baz 2> 'stderr.txt' 
cat 'stderr.txt'
```

### Character substitution with `tr`  (transliteration)

#### Switch case.


```{bash}
echo "gattaca" | tr a-z A-Z
```

#### Find reverse complement of DNA string.


```{bash}
echo 'GATTACA' | tr ACTG TGAC | rev
```

# Working with text

## Create a FASTA file to play with


```{bash}
cat > bgp.fasta << EOF
>HSBGPG Human gene for bone gla protein (BGP)
GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
ATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACC
ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
GAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAAC
CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG
GCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTC
CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC
CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
GGCCTATCGGCGCTTCTACGGCCCGGTCTAGGGTGTCGCTCTGCTGGCCTGGCCGGCAACCCCAGTTCTGCTCCT
CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC
ATCCCAGCTGCTCCCAAATAAACTCCAGAAG
EOF
```


```{bash}
wc bgp.fasta
```

## Using regular expresssions


```{bash}
cat bgp.fasta | 
grep "CCCCC"
```

```{bash}
cat bgp.fasta | 
grep -n "CCCCC"
```

```{bash}
cat bgp.fasta |
grep -nv "CCCCC"
```

```{bash}
cat bgp.fasta |
grep "GA*TT.*CA"
```

```{bash}
cat bgp.fasta |
grep -E "^C"
```

```{bash}
cat bgp.fasta |
grep -E "G$"
```

```{bash}
cat bgp.fasta |
grep -E "^C.*G$"
```

```{bash}
cat bgp.fasta |
grep -o "GA*TT.*CA"
```

```{bash}
cat bgp.fasta | 
grep -E "(GCAT)+"
```

```{bash}
cat bgp.fasta | 
grep -Eon "(GCA){2,}"
```

## Transliteration


```{bash}
cat bgp.fasta |
grep -E "^C.*G$" 
```

### Complement


```{bash}
cat bgp.fasta |
grep -E "^C.*G$" |
tr ACTG TAGC
```

### Reverse complement


```{bash}
cat bgp.fasta |
grep -E "^C.*G$" |
tr ACTG TAGC |
rev
```

## Sorting


```{bash}
cat bgp.fasta |
grep -v "^>" |
sort
```

### Sort by default uses lexicographic order


```{bash}
cat bgp.fasta |
grep -nv "^>" |
sort
```

### use `-n` flag for numeric order


```{bash}
cat bgp.fasta |
grep -nv "^>" |
sort -n
```

### Sort descending


```{bash}
cat bgp.fasta |
grep -nv "^>" |
sort -rn
```

## Downloading files


```{bash}
wget ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
```

```{bash}
ls
```

## File compression/uncompression


```{bash}
ls -lh Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
```

```{bash}
gunzip Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
```


```{bash}
ls -lh
```

## Inspecting the GTF file

A GTF file has some header lines, followed by tabular data in 9 columns:

```
chromosome name > chr{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y,M}
annotation source > {ENSEMBL,HAVANA}
feature-type > {gene,transcript,exon,CDS,UTR,start_codon,stop_codon,Selenocysteine}
genomic start location > integer-value (1-based)
genomic end location > integer-value
score (not used) > .
genomic strand > {+,-}
genomic phase (for CDS features) > {0,1,2,.}
additional information as key-value pairs > (format: key “value”;)
```


```{bash}
head Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf
```

```{bash}
tail Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf
```

## Remove comment lines

#### If you know the number of lines


```{bash}
tail +6 Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_no_comment.gtf 
head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_no_comment.gtf  
```

#### Using regular expressions (advanced)


```{bash}
cat Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf | grep -v '^#' > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_no_comment.gtf 

head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_no_comment.gtf 
```

## Spliting columns


```{bash}
cat Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf |
grep -v '^#' |
cut -f3 > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf 
head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
```

    gene
    transcript
    exon



```{bash}
cat Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf |
grep -v '^#' |
cut -f4-5 > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
```

    100	5645
    100	5645
    5494	5645



```{bash}
cat Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf |
grep -v '^#' |
cut -f2,4-5 > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
```

    ena	100	5645
    ena	100	5645
    ena	5494	5645


## Exercises

1. What is the mRNA version of bgp.fasta?


```{bash}

```

2. Extract the nucleotides in positions 5,10 and 15 of each line of bgp.fasta.


```{bash}

```

3. Find the number of mitochondrial exons in the GTF file.


```{bash}

```
