# The Unix Shell: Working with Text

## Text streams

Input and output of most Unix shell programs consists of plain text streams. Text output from a program can be piped into another program, or redirected to other streams. The standard streams are `stdin (0)` (standard input), `stdout (1)` (standard output) and `stderr (2)` (standard error). The default is to assume that input comes from `stdin` and output goes to `stdout`. We can also stream to and from a file. 

## Pipes and redirection

### Creating a text file from command line

For simple file creation, we can just use re-direction

A single `>` will create a new file or over-write an existing one.


```bash
echo "1 Hello, bash" > hello.txt
```


```bash
ls *txt
```

    hello.txt  stderr.txt  stdout.txt


### Appending


```bash
echo "2 Hello, again" >> hello.txt
```

#### Special non-printing characters


```bash
echo -e "3 Hello\n4 again" >> hello.txt
```

### From file to `stdout


```bash
cat hello.txt
```

    1 Hello, bash
    2 Hello, again
    3 Hello
    4 again


### Pipe to `cut` program to extract columns 2,3,4,5


```bash
cat hello.txt | cut -c 2-5
```

     Hel
     Hel
     Hel
     aga


### Building a chain of pipes

`wc -lc` reports the number of lines and bytes (usually corresponds to characters when using English text). 

Note that character count is 5 per line and not 4 because cut adds a newline character for each line.

#### Extract columns 2-5 and then count the number of lines and characters 


```bash
cat hello.txt | cut -c 2-5 | wc -lc  
```

          4      20


#### Extract lines 2-3


```bash
cat hello.txt | head -n 3 | tail -n 2
```

    2 Hello, again
    3 Hello


### Capturing error messages

The redirection operator > is actually 1> - that is, using stdout. We can also use 2> to redirect the output of stderr. &> means redirect both stdout and stderr, and is useful if for example, you want to direct all output to the same log file for later inspection.


```bash
mkdir foo/bar/baz > 'stdout.txt'
```

    mkdir: cannot create directory ‘foo/bar/baz’: No such file or directory




#### As there is nothing from stdout the file is empty


```bash
cat 'stdout.txt'
```

#### We need to use 2> to capture the output from stderr


```bash
mkdir foo/bar/baz 2> 'stderr.txt' 
cat 'stderr.txt'
```

    mkdir: cannot create directory ‘foo/bar/baz’: No such file or directory


### Character substitution with `tr`  (transliteration)

#### Switch case.


```bash
echo "gattaca" | tr a-z A-Z
```

    GATTACA


#### Find reverse complement of DNA string.


```bash
echo 'GATTACA' | tr ACTG TGAC | rev
```

    TGTAATC


# Working with text

## Create a FASTA file to play with


```bash
cat > bgp.fasta << EOF
>HSBGPG Human gene for bone gla protein (BGP)
GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
ATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACC
ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
GAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAAC
CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG
GCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTC
CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC
CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
GGCCTATCGGCGCTTCTACGGCCCGGTCTAGGGTGTCGCTCTGCTGGCCTGGCCGGCAACCCCAGTTCTGCTCCT
CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC
ATCCCAGCTGCTCCCAAATAAACTCCAGAAG
EOF
```


```bash
wc bgp.fasta
```

      18   25 1294 bgp.fasta


## Using regular expresssions


```bash
cat bgp.fasta | 
grep "CCCCC"
```

    GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
    TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG



```bash
cat bgp.fasta | 
grep -n "CCCCC"
```

    2:GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
    9:TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG



```bash
cat bgp.fasta |
grep -nv "CCCCC"
```

    1:>HSBGPG Human gene for bone gla protein (BGP)
    3:ATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACC
    4:ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
    5:CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    6:GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
    7:GAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAAC
    8:CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
    10:GCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTC
    11:CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
    12:GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
    13:GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
    14:AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC
    15:CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
    16:GGCCTATCGGCGCTTCTACGGCCCGGTCTAGGGTGTCGCTCTGCTGGCCTGGCCGGCAACCCCAGTTCTGCTCCT
    17:CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC
    18:ATCCCAGCTGCTCCCAAATAAACTCCAGAAG



```bash
cat bgp.fasta |
grep "GA*TT.*CA"
```

    GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
    GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
    TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG
    GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
    CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA



```bash
cat bgp.fasta |
grep -E "^C"
```

    CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
    CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
    CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
    CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC



```bash
cat bgp.fasta |
grep -E "G$"
```

    CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
    TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG
    CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
    GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
    GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
    ATCCCAGCTGCTCCCAAATAAACTCCAGAAG



```bash
cat bgp.fasta |
grep -E "^C.*G$"
```

    CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG



```bash
cat bgp.fasta |
grep -o "GA*TT.*CA"
```

    GATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCA
    GTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCA
    GTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCA
    GTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCA
    GTTGGCTGACCACATCGGCTTTCA



```bash
cat bgp.fasta | 
grep -E "(GCAT)+"
```

    GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
    ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
    CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC



```bash
cat bgp.fasta | 
grep -Eon "(GCA){2,}"
```

    3:GCAGCA
    12:GCAGCA


## Transliteration


```bash
cat bgp.fasta |
grep -E "^C.*G$" 
```

    CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG


### Complement


```bash
cat bgp.fasta |
grep -E "^C.*G$" |
tr ACTG TAGC
```

    ATAAGAAAAGATCCAACATGGCATCGCCCCCAGCTCTCCTCCTTCATAATGCCAAATAAGAGGAGATAAAAGGGC
    ATATCAAGGGCGCGAATTCATCCTCCCATCACTCCGTCGCTTCTCTAAATCCACAGTAAGCGTGATTGCCAGCCC


### Reverse complement


```bash
cat bgp.fasta |
grep -E "^C.*G$" |
tr ACTG TAGC |
rev
```

    CGGGAAAATAGAGGAGAATAAACCGTAATACTTCCTCCTCTCGACCCCCGCTACGGTACAACCTAGAAAAGAATA
    CCCGACCGTTAGTGCGAATGACACCTAAATCTCTTCGCTGCCTCACTACCCTCCTACTTAAGCGCGGGAACTATA


## Sorting


```bash
cat bgp.fasta |
grep -v "^>" |
sort
```

    AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC
    ATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACC
    ATCCCAGCTGCTCCCAAATAAACTCCAGAAG
    ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
    CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
    CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
    CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
    CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC
    GAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAAC
    GCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTC
    GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
    GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
    GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
    GGCCTATCGGCGCTTCTACGGCCCGGTCTAGGGTGTCGCTCTGCTGGCCTGGCCGGCAACCCCAGTTCTGCTCCT
    GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
    TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG


### Sort by default uses lexicographic order


```bash
cat bgp.fasta |
grep -nv "^>" |
sort
```

    10:GCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTC
    11:CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
    12:GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
    13:GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
    14:AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC
    15:CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
    16:GGCCTATCGGCGCTTCTACGGCCCGGTCTAGGGTGTCGCTCTGCTGGCCTGGCCGGCAACCCCAGTTCTGCTCCT
    17:CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC
    18:ATCCCAGCTGCTCCCAAATAAACTCCAGAAG
    2:GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
    3:ATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACC
    4:ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
    5:CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    6:GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
    7:GAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAAC
    8:CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
    9:TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG


### use `-n` flag for numeric order


```bash
cat bgp.fasta |
grep -nv "^>" |
sort -n
```

    2:GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
    3:ATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACC
    4:ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
    5:CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    6:GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
    7:GAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAAC
    8:CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
    9:TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG
    10:GCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTC
    11:CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
    12:GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
    13:GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
    14:AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC
    15:CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
    16:GGCCTATCGGCGCTTCTACGGCCCGGTCTAGGGTGTCGCTCTGCTGGCCTGGCCGGCAACCCCAGTTCTGCTCCT
    17:CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC
    18:ATCCCAGCTGCTCCCAAATAAACTCCAGAAG


### Sort descending


```bash
cat bgp.fasta |
grep -nv "^>" |
sort -rn
```

    18:ATCCCAGCTGCTCCCAAATAAACTCCAGAAG
    17:CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC
    16:GGCCTATCGGCGCTTCTACGGCCCGGTCTAGGGTGTCGCTCTGCTGGCCTGGCCGGCAACCCCAGTTCTGCTCCT
    15:CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
    14:AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC
    13:GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
    12:GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
    11:CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
    10:GCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTC
    9:TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG
    8:CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
    7:GAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAAC
    6:GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
    5:CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
    4:ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
    3:ATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACC
    2:GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT


## Downloading files


```bash
wget ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
```

    --2020-06-22 19:24:55--  ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
               => ‘Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz.1’
    Resolving ftp.ensemblgenomes.org (ftp.ensemblgenomes.org)... 193.62.193.141
    Connecting to ftp.ensemblgenomes.org (ftp.ensemblgenomes.org)|193.62.193.141|:21... connected.
    Logging in as anonymous ... Logged in!
    ==> SYST ... done.    ==> PWD ... done.
    ==> TYPE I ... done.  ==> CWD (1) /pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99 ... done.
    ==> SIZE Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz ... 1796344
    ==> PASV ... done.    ==> RETR Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz ... done.
    Length: 1796344 (1.7M) (unauthoritative)
    
    Cryptococcus_neofor 100%[===================>]   1.71M  3.09MB/s    in 0.6s    
    
    2020-06-22 19:24:57 (3.09 MB/s) - ‘Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz.1’ saved [1796344]
    



```bash
ls
```

    1_IntroToCompBoot.ipynb
    bgp.fasta
    [0m[01;31mCryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz[0m
    Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz.1
    [01;34mfigs[0m
    hello.txt
    R01_IntroductionToR.ipynb
    R02_Dplyr.ipynb
    R03_Tidyr.ipynb
    R04_Multiple_Files.ipynb
    R05_Graphics_Overview.ipynb
    R06_Graphics_Base.ipynb
    R07_Tidyverse_With_Pilot_Counts.ipynb
    R07_Tidyverse_With_Pilot_Counts_Solutions.ipynb
    R08_Graphics_ggplot2.ipynb
    R09_Libraries_and_Bioconductor.ipynb
    stderr.txt
    stdout.txt
    Unix01_Intro_to_Unix_and_bash.ipynb
    Unix02_Files_And_Directories.ipynb
    Unix03_File_Storage.ipynb
    Unix04_FileIO.ipynb
    Unix05_Text_Manipulation.ipynb
    Unix06_Variables.ipynb
    Unix07_Regular_Expresssions.ipynb
    Unix08_Capstone_Exercises.ipynb
    Unix08_Capstone_Exercises_Solutions.ipynb


## File compression/uncompression


```bash
ls -lh Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
```

    -rw-r--r-- 1 jovyan users 1.8M Jun 22 19:14 [0m[01;31mCryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz[0m



```bash
gunzip Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
```


```bash
ls -lh
```

    total 36M
    -rw-r--r-- 1 jovyan users 5.5K Feb 19 17:49 1_IntroToCompBoot.ipynb
    -rw-r--r-- 1 jovyan users 1.3K Jun 22 19:24 bgp.fasta
    -rw-r--r-- 1 jovyan users  32M Jun 22 19:14 Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf
    -rw-r--r-- 1 jovyan users 1.8M Jun 22 19:24 Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz.1
    drwxr-xr-x 2 jovyan users 4.0K Jun 22 17:57 [0m[01;34mfigs[0m
    -rw-r--r-- 1 jovyan users   45 Jun 22 19:24 hello.txt
    -rw-r--r-- 1 jovyan users  65K Jun 20 21:06 R01_IntroductionToR.ipynb
    -rw-r--r-- 1 jovyan users 116K Jun 20 21:14 R02_Dplyr.ipynb
    -rw-r--r-- 1 jovyan users  86K Jun 22 16:50 R03_Tidyr.ipynb
    -rw-r--r-- 1 jovyan users 121K Jun 22 16:51 R04_Multiple_Files.ipynb
    -rw-r--r-- 1 jovyan users 155K Jun 22 17:24 R05_Graphics_Overview.ipynb
    -rw-r--r-- 1 jovyan users 6.0K Jun 22 17:24 R06_Graphics_Base.ipynb
    -rw-r--r-- 1 jovyan users 141K Jun 22 17:42 R07_Tidyverse_With_Pilot_Counts.ipynb
    -rw-r--r-- 1 jovyan users 152K Jun 22 17:43 R07_Tidyverse_With_Pilot_Counts_Solutions.ipynb
    -rw-r--r-- 1 jovyan users 1.4M Jun 22 18:03 R08_Graphics_ggplot2.ipynb
    -rw-r--r-- 1 jovyan users 353K Jun 22 18:42 R09_Libraries_and_Bioconductor.ipynb
    -rw-r--r-- 1 jovyan users   76 Jun 22 19:24 stderr.txt
    -rw-r--r-- 1 jovyan users    0 Jun 22 19:24 stdout.txt
    -rw-r--r-- 1 jovyan users 8.6K Jun 22 18:43 Unix01_Intro_to_Unix_and_bash.ipynb
    -rw-r--r-- 1 jovyan users  27K Jun 22 18:46 Unix02_Files_And_Directories.ipynb
    -rw-r--r-- 1 jovyan users  16K Jun 22 18:47 Unix03_File_Storage.ipynb
    -rw-r--r-- 1 jovyan users  24K Jun 22 18:51 Unix04_FileIO.ipynb
    -rw-r--r-- 1 jovyan users  35K Jun 22 19:24 Unix05_Text_Manipulation.ipynb
    -rw-r--r-- 1 jovyan users  16K Feb 19 17:49 Unix06_Variables.ipynb
    -rw-r--r-- 1 jovyan users  11K Feb 19 17:49 Unix07_Regular_Expresssions.ipynb
    -rw-r--r-- 1 jovyan users 5.3K Feb 19 17:49 Unix08_Capstone_Exercises.ipynb
    -rw-r--r-- 1 jovyan users 8.1K Feb 19 17:49 Unix08_Capstone_Exercises_Solutions.ipynb


## Inspecting the GTF file

A GTF file has some header lines, followed by tabular data in 9 columns:

```
chromosome name > chr{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y,M}
annotation source > {ENSEMBL,HAVANA}
feature-type > {gene,transcript,exon,CDS,UTR,start_codon,stop_codon,Selenocysteine}
genomic start location > integer-value (1-based)
genomic end location > integer-value
score (not used) > .
genomic strand > {+,-}
genomic phase (for CDS features) > {0,1,2,.}
additional information as key-value pairs > (format: key “value”;)
```


```bash
head Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf
```

    #!genome-build CNA3
    #!genome-version CNA3
    #!genome-date 2015-11
    #!genome-build-accession GCA_000149245.3
    #!genebuild-last-updated 2015-11
    1	ena	gene	100	5645	.	-	.	gene_id "CNAG_04548"; gene_source "ena"; gene_biotype "protein_coding";
    1	ena	transcript	100	5645	.	-	.	gene_id "CNAG_04548"; transcript_id "AFR92135"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding";
    1	ena	exon	5494	5645	.	-	.	gene_id "CNAG_04548"; transcript_id "AFR92135"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding"; exon_id "AFR92135-1";
    1	ena	CDS	5494	5645	.	-	0	gene_id "CNAG_04548"; transcript_id "AFR92135"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding"; protein_id "AFR92135"; protein_version "1";
    1	ena	start_codon	5643	5645	.	-	0	gene_id "CNAG_04548"; transcript_id "AFR92135"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding";



```bash
tail Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf
```

    Mt	ena	stop_codon	23840	23842	.	+	0	gene_id "CNAG_09011"; transcript_id "AFR99113"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding";
    Mt	Ensembl_Fungi	gene	23909	23980	.	+	.	gene_id "ENSRNA049545749"; gene_name "tRNA-Val"; gene_source "Ensembl_Fungi"; gene_biotype "tRNA";
    Mt	Ensembl_Fungi	transcript	23909	23980	.	+	.	gene_id "ENSRNA049545749"; transcript_id "ENSRNA049545749-T1"; gene_name "tRNA-Val"; gene_source "Ensembl_Fungi"; gene_biotype "tRNA"; transcript_source "Ensembl_Fungi"; transcript_biotype "tRNA";
    Mt	Ensembl_Fungi	exon	23909	23980	.	+	.	gene_id "ENSRNA049545749"; transcript_id "ENSRNA049545749-T1"; exon_number "1"; gene_name "tRNA-Val"; gene_source "Ensembl_Fungi"; gene_biotype "tRNA"; transcript_source "Ensembl_Fungi"; transcript_biotype "tRNA"; exon_id "ENSRNA049545749-E1";
    Mt	ena	gene	24096	24851	.	+	.	gene_id "CNAG_09012"; gene_source "ena"; gene_biotype "protein_coding";
    Mt	ena	transcript	24096	24851	.	+	.	gene_id "CNAG_09012"; transcript_id "AFR99114"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding";
    Mt	ena	exon	24096	24851	.	+	.	gene_id "CNAG_09012"; transcript_id "AFR99114"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding"; exon_id "AFR99114-1";
    Mt	ena	CDS	24096	24848	.	+	0	gene_id "CNAG_09012"; transcript_id "AFR99114"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding"; protein_id "AFR99114"; protein_version "1";
    Mt	ena	start_codon	24096	24098	.	+	0	gene_id "CNAG_09012"; transcript_id "AFR99114"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding";
    Mt	ena	stop_codon	24849	24851	.	+	0	gene_id "CNAG_09012"; transcript_id "AFR99114"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding";


## Remove comment lines

#### If you know the number of lines


```bash
tail +6 Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_no_comment.gtf 
head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_no_comment.gtf  
```

    1	ena	gene	100	5645	.	-	.	gene_id "CNAG_04548"; gene_source "ena"; gene_biotype "protein_coding";
    1	ena	transcript	100	5645	.	-	.	gene_id "CNAG_04548"; transcript_id "AFR92135"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding";
    1	ena	exon	5494	5645	.	-	.	gene_id "CNAG_04548"; transcript_id "AFR92135"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding"; exon_id "AFR92135-1";


#### Using regular expressions (advanced)


```bash
cat Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf | grep -v '^#' > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_no_comment.gtf 

head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_no_comment.gtf 
```

    1	ena	gene	100	5645	.	-	.	gene_id "CNAG_04548"; gene_source "ena"; gene_biotype "protein_coding";
    1	ena	transcript	100	5645	.	-	.	gene_id "CNAG_04548"; transcript_id "AFR92135"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding";
    1	ena	exon	5494	5645	.	-	.	gene_id "CNAG_04548"; transcript_id "AFR92135"; exon_number "1"; gene_source "ena"; gene_biotype "protein_coding"; transcript_source "ena"; transcript_biotype "protein_coding"; exon_id "AFR92135-1";


## Spliting columns


```bash
cat Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf |
grep -v '^#' |
cut -f3 > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf 
head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
```

    gene
    transcript
    exon



```bash
cat Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf |
grep -v '^#' |
cut -f4-5 > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
```

    100	5645
    100	5645
    5494	5645



```bash
cat Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf |
grep -v '^#' |
cut -f2,4-5 > Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
head -3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39_split.gtf
```

    ena	100	5645
    ena	100	5645
    ena	5494	5645


## Exercises

1. Waht is the mRNA version of bgp.fasta?


```bash

```

2. Extract the nucleotides in positions 5,10 and 15 of each line of bgp.fasta.


```bash

```

3. Find the number of mitochondrial exons in the GTF file.


```bash

```
