---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.3
  kernelspec:
    display_name: Bash
    language: bash
    name: bash
---

<!-- #region -->
# Making an adapter file


## Shell Variables
Retyping shell variables in every notebook is gets old, and it is error prone.  Let's centralize these so we can share them between notebooks.  We can create a shell script that contains the shell variables that we need, and then we can `source` it in each notebook.  Let's call it `bioinf_config.sh`.
<!-- #endregion -->

```{bash}
source bioinf_config.sh
```

## Making New Directories
Make the directories that are new in this notebook

```{bash}
mkdir -p $OUT_DIR
```

## Making an adapter file
As part of trimming and filtering we want to remove adapter contamination. To do this we need to make a file containing the adapter sequences that we can supply to our trimming software.

The first step is find out what the adapter sequences are! To do this, we need to know what adapters were used to make our libraries. If we made the libraries ourselves, then we should know. If the libraries were made by a sequencing core, then it is best to get this information from them at the time of sequencing. Sometimes the core can give us the adapter sequences and sometimes they will just tell us the kit information and we need to look up the sequences ourselves.

In our case, the [paper](https://pubmed.ncbi.nlm.nih.gov/33597266/) doesn't tell us anything about the library prep, but it links to [Supplementary Methods](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8130008/bin/NIHMS1696940-supplement-Supplemental_material_Fin.docx) which tells us that the TruSeq RNA Sample Preparation Kit v2 was used for library prep. So we can get the [adapter sequences from Illumina](https://support-docs.illumina.com/SHARE/AdapterSeq/Content/SHARE/AdapterSeq/TruSeq/CDIndexes.htm).

So the adapter sequences are:

- truseq_read_1: AGATCGGAAGAGCACACGTCTGAACTCCAGTCA
- truseq_read_2: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT
    
I like adding the reverse complement of the adapter sequences because it is easier than figuring out how they will come off the sequencer.  There are lots of ways to generate the reverse complement.  One of the easiest is a web based tool: https://www.bioinformatics.org/sms/rev_comp.html, which also gives us:

- truseq_read_1_rc: TGACTGGAGTTCAGACGTGTGCTCTTCCGATCT
- truseq_read_2_rc: ACACTCTTTCCCTACACGACGCTCTTCCGATCT

We can make the adapter file with the Jupyter text editor, but it might be easier to use echo and redirect it to a file. Most trimming sofware wants FASTA format, so it should look like this:

```{bash}
# https://support-docs.illumina.com/SHARE/AdapterSeq/Content/SHARE/AdapterSeq/TruSeq/CDIndexes.htm
echo ">truseq_read_1
AGATCGGAAGAGCACACGTCTGAACTCCAGTCA
>truseq_read_2
AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT
>truseq_read_1_rc
TGACTGGAGTTCAGACGTGTGCTCTTCCGATCT
>truseq_read_2_rc
ACACTCTTTCCCTACACGACGCTCTTCCGATCT
" > $ADAPTER_FASTA
```

```{bash}
cat $ADAPTER_FASTA
```
