---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.3
  kernelspec:
    display_name: Bash
    language: bash
    name: bash
---

# Mapping Reads to a Reference Genome
<!---
** Add Intro Text***
-->

## Shell Variables

```{bash}
# Source the config script
source bioinf_config.sh

mkdir -p $STAROUT_DIR
ls $OUT_DIR
```

## Mapping with STAR


```{bash}
STAR \
    --runMode alignReads \
    --twopassMode None \
    --genomeDir $GENOME_DIR \
    --readFilesIn $TRIM_DIR/SRR12804466_1_trimmed.fastq.gz \
    --outFileNamePrefix ${STAROUT_DIR}/SRR12804466_1_ \
    --readFilesCommand gunzip -c \
    --quantMode GeneCounts \
    --outSAMtype BAM SortedByCoordinate \
    --runThreadN $TOTAL_THREADS
```

We will start with these parameters, but there is an extensive list of command line options detailed in the [STAR Manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf), it is a good idea to read through and try to understand all of them.  We will discuss some more later.

* --runMode alignReads : map reads 
* --twopassMode : Run one pass or two? If two-pass mode is on, STAR tries to discover novel junctions, then reruns mapping with these added to the annotation
* --genomeDir : directory containing the genome index
* --readFilesIn : input FASTQ
* --readFilesCommand gunzip -c : use "gunzip -c" to uncompress FASTQ on-the-fly, since it is gzipped
* --outFileNamePrefix : prefix (and path) to use for all output files
* --quantMode GeneCounts : output a table of read counts per gene
* --outSAMtype BAM Unsorted : output an unsort BAM file
* --outSAMunmapped Within : included unmapped reads in the BAM file
* --runThreadN : tells STAR to run using multiple cores.  I am using it so we don't have to wait too long for this to run during class.  It is OK to use multiple cores, but before you do this you should be sure that the server is not busy, and even then you should use a reasonable number of cores.  Abusing multi-threading is inconsiderate of other users and could crash the server.



### STAR Output
So what happened? Let's take a look . . .

```{bash}
ls ${STAROUT_DIR}
```

```{bash}
head ${STAROUT_DIR}/SRR12804466_1*.out ${STAROUT_DIR}/SRR12804466_1*.tab
```

STAR generates several files for each FASTQ:
* Log.out : lots of details of the run, including all parameters used
* Log.final.out : Important summary statistics
* ReadsPerGene.out.tab : Count table, the main thing we are interested in
* SJ.out.tab : All splice junctions, including ones from the GTF and novel junctions discovered by STAR
* Log.progress.out: run statistics updated during run, not so interesting at the end

Let's take a closer look at Log.final.out

```{bash}
cat ${STAROUT_DIR}/SRR12804466_1_Log.final.out
```

### Reality Checks

#### Reality Check: Number of Reads
Are the number of input reads what we expect? Let's look at how many reads are in the input FASTQ.  We can use MultiQC to look at this and other aspects of the mapping results.

```{bash}
multiqc ${QC} ${STAROUT_DIR} --outdir ${STAROUT_DIR}
```

Once multiqc is done running we can view the results by finding the output in the Jupyter browser, it should be in a file named `multiqc_report.html` in the following directory.

It is 
We should look to Are the number of input reads what we expect? Let's look at how many reads are in the input FASTQ

```{bash}
echo ${STAROUT_DIR}
```

#### Reality Check: Unmapped Reads
It is always a good idea to examine a sample of unmapped reads to figure out what they are. The easiest way to do this is with BLAST.  In the past I have discovered that an experiment was contaminated with a different species by BLASTing unmapped reads.  In that case there were a large number of unmapped reads, which raised my suspicions.

Even with a high rate of mapped reads, it is worth spending a few minutes to check them out.  The simplest thing to do is use `samtools` to generate a FASTA from the unmapped.bam, grab a few of these sequences, and then [BLAST them against the nr database](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome)

```{bash}
samtools view -f 4 ${STAROUT_DIR}/SRR12804466_1_Aligned.sortedByCoord.out.bam

```

```{bash}
samtools fasta -f 4 ${STAROUT_DIR}/SRR12804466_1_Aligned.sortedByCoord.out.bam | head -n20
```

#### Reality Check: Strandness

```{bash}
infer_experiment.py -r ${BED_PATH} -i ${STAROUT_DIR}/SRR12804466_1_Aligned.sortedByCoord.out.bam
```

# Picard CollectRnaSeqMetrics

`PicardCommandLine CollectRnaSeqMetrics I=/star_output/Aligned.sortedByCoord.out.bam O=/picard_output/RNA_Metrics.txt REF_FLAT=genome.refFlat STRAND=FIRST_READ_TRANSCRIPTION_STRAND RIBOSOMAL_INTERVALS=genome.intervalList`
