---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.3
  kernelspec:
    display_name: Bash
    language: bash
    name: bash
---

# Mapping Reads to a Reference Genome
Mapping paired-end reads is only slightly more complicated than mapping single end reads.

Let's load variables from bioinf_config.sh, make a new directory for output, and check that it looks OK.

```{bash}
# Source the config script
source bioinf_config.sh

mkdir -p ${PE_STAROUT_DIR}
ls $PE_OUT_DIR
```

## Mapping with STAR
The only thing we need to change for paired-end reads is that we supply STAR with both the foward and reverse FASTQs


We will start with these parameters, but there is an extensive list of command line options detailed in the [STAR Manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf), it is a good idea to read through and try to understand all of them.  We will discuss some more later.

* --runMode alignReads : map reads 
* --twopassMode : Run one pass or two? If two-pass mode is on, STAR tries to discover novel junctions, then reruns mapping with these added to the annotation
* --genomeDir : directory containing the genome index
* --readFilesIn : input forward FASTQ and input reverse FASTQ
* --readFilesCommand gunzip -c : use "gunzip -c" to uncompress FASTQ on-the-fly, since it is gzipped
* --outFileNamePrefix : prefix (and path) to use for all output files
* --quantMode GeneCounts : output a table of read counts per gene
* --outSAMtype BAM Unsorted : output an unsort BAM file
* --outSAMunmapped Within : included unmapped reads in the BAM file
* --runThreadN : tells STAR to run using multiple cores.  I am using it so we don't have to wait too long for this to run during class.  It is OK to use multiple cores, but before you do this you should be sure that the server is not busy, and even then you should use a reasonable number of cores.  Abusing multi-threading is inconsiderate of other users and could crash the server.


```{bash}
STAR \
    --runMode alignReads \
    --twopassMode None \
    --genomeDir $GENOME_DIR \
    --readFilesIn ${PE_TRIM_DIR}/SRR12804466_1_trimmed.fastq.gz ${PE_TRIM_DIR}/SRR12804466_2_trimmed.fastq.gz \
    --readFilesCommand gunzip -c \
    --outFileNamePrefix ${PE_STAROUT_DIR}/SRR12804466_ \
    --quantMode GeneCounts \
    --outSAMtype BAM SortedByCoordinate \
    --outSAMunmapped Within \
    --runThreadN $TOTAL_THREADS
```

### STAR Output
So what happened? Let's take a look . . .

```{bash}
ls ${PE_STAROUT_DIR}
```

```{bash}
head ${PE_STAROUT_DIR}/SRR12804466_*.out ${PE_STAROUT_DIR}/SRR12804466_*.tab
```

STAR generates several files for each FASTQ:
* Log.out : lots of details of the run, including all parameters used
* Log.final.out : Important summary statistics
* ReadsPerGene.out.tab : Count table, the main thing we are interested in
* SJ.out.tab : All splice junctions, including ones from the GTF and novel junctions discovered by STAR
* Log.progress.out: run statistics updated during run, not so interesting at the end

Let's take a closer look at Log.final.out

```{bash}
cat ${PE_STAROUT_DIR}/SRR12804466_Log.final.out
```

### Summary Statistics 

Let's check out the summary statistics in MultiQC and remember to compare the number of input (raw) reads with the number of mapped reads.

```{bash}
multiqc --force $PE_QC $PE_STAROUT_DIR --outdir $PE_QC
```

Once multiqc is done running we can view the results by finding the output in the Jupyter browser, it should be in a file named `multiqc_report.html` in the following directory.

```{bash}
echo ${STAROUT_DIR}
```

