---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.2
  kernelspec:
    display_name: R
    language: R
    name: ir
---

# Import annotation from a GTF or GFF file

```{r}
library(tidyverse)
library(rtracklayer)
library(DESeq2)
library(plyranges)
```

```{r}
### Import dds analysis file
procdir <- normalizePath("./proc/")
ddsfile <- file.path(procdir, "PRJNA668393-ddsdata.RDS")

PRJNA668393ddsData <- readRDS(ddsfile)

tools::md5sum(ddsfile)
```

```{r}
annodir <- normalizePath("/data/genome/")

gfffile <- file.path(annodir, "gencode.vM27.primary_assembly.annotation.gff3")
tools::md5sum(gfffile)

gencodeidfile <- "gencode.vM27.metadata.EntrezGene.gz"

```

```{r}
# Import gff file
### Import gtf file
rtracklayer::import(gfffile) ->
    gffdat
```

```{r}
## Inspect imported object
gffdat
```

Filter by gene name

```{r}
gffdat %>% plyranges::filter(gene_name == "Il17f") %>% tibble::as_tibble() %>% pull(Parent)
```

Filter by gene symbol and exons

```{r}
gffdat %>% plyranges::filter(gene_name == "Il17f" & type == "exon") %>% tibble::as_tibble()
```

## Import ensembl id to entrez id file from gencode

```{r}
readr::read_tsv(gencodeidfile, col_names = c("ensid", "entrez"), col_types = c(col_character(), col_character())) -> 
    idmapdat
```

```{r}
idmapdat
```

```{r}
gffdat %>% 
    plyranges::filter(type == "transcript") %>% 
    as_tibble() %>% 
    dplyr::select(gene_id, transcript_id, gene_name, mgi_id) %>% 
    dplyr::full_join(idmapdat, by = c("transcript_id" = "ensid")) 
```

```{r}
%>%
    dplyr::distinct(gene_id, entrez) %>%
    dplyr::full_join(tibble(gid = rownames(PRJNA668393ddsData)), by = c("gene_id" = "gid")) -> 
    idmap
```

```{r}
idmap %>% drop_na
```

```{r}

```

## Extract select annotation from the object as a tibble

```{r}
gffdat %>%
    plyranges::filter(type == "gene") %>%
    tibble::as_tibble() %>%
    dplyr::select(seqnames, start, end, strand, gene_type, mgi_id, gene_id, gene_name) ->
    ens2symbdat
```

```{r}
ens2symbdat
```

## Check if the ensembl ids agree with the featureNames in the DESeq2 object

```{r}
setequal(rownames(PRJNA668393ddsData), ens2symbdat[["gene_id"]])
```

## Examine unqiueness of ensembl ids and symbols

```{r}
### ensembl ids are unique
ens2symbdat %>% dplyr::pull(gene_id) %>% duplicated() %>% any
```

```{r}
### gene symbols are not
ens2symbdat %>% dplyr::pull(gene_name) %>% duplicated() %>% any
```

```{r}
### duplicate gene names
ens2symbdat %>% 
    dplyr::group_by(gene_name) %>% 
    dplyr::filter(n()>1)
```

```{r}
### focusing on protein coding genes will reduce the number of duplicate gene ids
ens2symbdat %>% 
    dplyr::filter(gene_type == "protein_coding") %>% 
    dplyr::group_by(gene_name) %>% 
    dplyr::filter(n()>1)
```

```{r}
outdir <- "./proc/"

outfile <- file.path(outdir, "PRJNA668393-annodata.RDS")

saveRDS(ens2symbdat, file = outfile)

tools::md5sum(outfile)
```

```{r}
sessionInfo()
```

```{r}
gffdat %>% filter(type == "gene") %>% head(3) %>% as_tibble() %>% colnames()
```

```{r}
? rtracklayer::import
```

```{r}
wget http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/gencode.vM27.metadata.EntrezGene.gz
```

```{r}

```
