---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.2
  kernelspec:
    display_name: R
    language: R
    name: ir
---

# DESeq2: Basic Differential Expression (DE) analysis


## Objective: Carry out a basic set of DE analyses using DESeq2 and visualize the results


### Load packages

```{r}
library(tidyverse)
library(DESeq2)
library(dendextend)
library(RColorBrewer)
```

### Load the PRJNA668393 dds object from image file

```{r}
procdir <- normalizePath("./proc/")
ddsfile <- file.path(procdir, "PRJNA668393-ddsdata.RDS")

PRJNA668393ddsData <- readRDS(ddsfile)

tools::md5sum(ddsfile)
```

### Check dimensions of the two objects


# Inspect object & Slots of an S4 class


Let's has a look at the object we have created.

```{r}
PRJNA668393ddsData
```

see the class of dds object

```{r}
class(PRJNA668393ddsData)
```

DESeqDataSet is a S4 object. Recall that a S4 object was taught when introducing bioconductor. Note that S4 objects allow users to wrap up multiple elements into a single variables where each element is called a slot.

```{r}
slotNames(PRJNA668393ddsData)
```

The metadata (columnData) is stored in the slot `colData`

```{r}
colData(PRJNA668393ddsData) %>% as.data.frame
```

The design formula is stored in the slot `design`. The design holds the R formula which expresses how the counts depend on the variables in colData.

```{r}
design(PRJNA668393ddsData)
```

The first thing you may want to do is **have a look at the raw counts** you have imported. The `DESeq2::counts` function extracts a matrix of counts (with the genes along the rows and samples along the columns). Let us first verify the dimension of this matrix.

```{r}
dim(DESeq2::counts(PRJNA668393ddsData))
```

```{r}
head(DESeq2::counts(PRJNA668393ddsData),3)
```

This slot returns gene specific information (it will be populated later)

```{r}
dispersionFunction(PRJNA668393ddsData)
```

Get size factors

```{r}
sizeFactors(PRJNA668393ddsData)
```

# Differential expression analysis


## Specify Model


You recall that DESeq requires that  we have estimates for sample specific size factors and gene specific dispersion factors. More specifically, recall that DESeq models the count $K_{ij}$ (gene $i$, sample $j$) as negative binomial with mean $\mu_{ij}$ and dispersion parameter $\alpha_i$. Here $\mu_{ij}=s_j q_{ij}$ where $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} x_j + \beta_{2i} z_j$. Here $s_j$ is the sample $j$ specific size factor.

**Summarize of notation**
- $K_{ij}$ denotes the observed **number of reads** mapped to gene $i$ for sample $j$
- $K_{ij}$ follows a **negative binomial distribution** with
    - **Mean** $\mu_{ij}$
    - **Dispersion parameter** $\alpha_i$
- Modelling
    - $K_{ij} \sim NB(\mu_{ij}, \alpha_i)$
    - $\mu_{ij} = s_{j}q_{ij}$
        - $s_j$ is sample $j$ specific normalization constant
    - $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} x_j + \beta_{2i} z_j $
        - $x_j=0$ if treament = isotype, or
        - $x_j=1$ if treatment = anti-PD1
        - $z_j=0$ if cell line = Parental, or
        - $z_j=1$ if cell line = UV2.
 - Effect size (no cell line effect)
     - If there is differential treatment effect but NO cell line effect for gene $i$, then
         - $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i}$
     - The corresponding mean count is
         - $\mu_{ij} = s_j 2^{\beta_{0i} + \beta_{1i}}$
 
  - Effect size (with cell line effect)
     - If there is differential treatment and cell line effects for gene $i$, then
         - $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} + \beta_{2i}$
     - The corresponding mean count is
         - $\mu_{ij} = s_j 2^{\beta_{0i} + \beta_{1i} \beta_{2i}}$   


Add the design to the DESeq2 object

```{r}
ddsobj <- PRJNA668393ddsData
```

```{r}
design(ddsobj) <- formula(~ cline + trt)
```

```{r}
design(ddsobj)
```

## 01 Size Factors
 We begin by estimating the size factors $s_1,\ldots,s_n$:

```{r}
ddsobj <- estimateSizeFactors(ddsobj)
```

Now, compare the dds object to that of before applying the estimateSizeFactors() function. What has changed? What remains unchanged?

```{r}
ddsobj
```

Note that there is a **sizeFactor** added to **colData**. Let's look at it more carefully

You can also get the size factors directly

```{r}
tibble::tibble(sfactor = sizeFactors(ddsobj))
```

Summarize size factors

```{r}
summary(sizeFactors(ddsobj))
```

Now that the size factors have been estimated, we can get "normalized" counts (DESeq2 normalizes against size factor)

```{r}
# original counts for three genes for samples 3 and 11
counts(ddsobj)[102:104,c(3,11)]
```

```{r}
# Size factors for samples 3 and 11

sizeFactors(ddsobj)[c(3,11)]
```

```{r}
# normalized count for samples 3 and 11
counts(ddsobj, normalize = TRUE)[102:104, c(3,11)]
```

```{r}
# normalized manually using size factors for sample 3
counts(ddsobj)[102:104, 3] / sizeFactors(ddsobj)[3]
```

```{r}
# normalized manually using size factors for sample 11
counts(ddsobj)[102:104, 11] / sizeFactors(ddsobj)[11]
```

How do you get the raw and normalized counts for specific genes and samples

```{r}
counts(ddsobj, normalize = FALSE)[c("ENSMUSG00000061024.9", "ENSMUSG00000079671.9"),c("SRR12804465","SRR12804467")]
```

```{r}
counts(ddsobj, normalize = TRUE)[c("ENSMUSG00000061024.9", "ENSMUSG00000079671.9"),c("SRR12804465","SRR12804467")]
```

## 02 Dispersion Parameters
Next, we get the dispersion factors $\alpha_1,\ldots,\alpha_{m}$

```{r}
ddsobj <- estimateDispersions(ddsobj)
```

Now inspect the dds object again and note that the rowRanges slot has extra information ("metadata column names(0):" before versus "column names(9): baseMean baseVar ... dispOutlier dispMAP")
- before: 
    - `metadata column names(0):`
- after:  
    - `column names(9): baseMean baseVar ...`

```{r}
ddsobj
```

Can you notice the difference?



Note that the dispersionfunction slot is now populated

```{r}
dispersionFunction(ddsobj)
```

We can extract the gene specific dispersion factors using dispersions(). Note that there will be one number per gene. We look at the first four genes (rounded to 4 decimal places)

```{r}
alphas <- dispersions(ddsobj)
```

Verify that the number of dispersion factors equals the number of genes

```{r}
# number of disperion factors
length(alphas)
```

```{r}
summary(alphas)
```

Summarize the dispersion factors using a box plot (may want to log transform)

```{r}
boxplot(log(dispersions(ddsobj)))
```

# Differential Expression Analysis
We can now conduct a differential expression analysis using the DESeq() function. Keep in mind that to get to this step, we first estimated the size factors and then the dispersion parameters.

```{r}
### Carry out DE analysis
ddsDE <- DESeq2::DESeq(ddsobj)
```

```{r}
### Look at object
ddsDE
```

```{r}
### Look at some of the results
results(ddsDE)
```

### Look at some of the results (tidy version sorted by adjusted P-value)

```{r}
results(ddsDE, tidy = TRUE) %>% 
    dplyr::arrange(padj) %>%
        head(10)
```

We can get the results for the differential expression analysis using results(). Here, we can compare two group of samples specified by the contrast. (If not, the default contrast would be the last term in your additive model `design(dds)`).

```{r}
# DE with respect to treatment using isotype as the referent 
# (positive effect size suggests that treatment up-regulates)
DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"))
```

```{r}
# DE with respect to treatment using anti-PD-1 as the referent 
# (negative effect size suggests that treatment up-regulates)
DESeq2::results(ddsDE, contrast = c("trt", "isotype", "antiPD1"))
```

Let's look at the results for the first four genes

```{r}
### Tidy the results
DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>% 
    dplyr::arrange(padj) %>%
        head(10)
```

```{r}
### DE effect with respect to cell line
### Tidy the results
DESeq2::results(ddsDE, contrast = c("cline", "UV2", "Parental"), tidy = TRUE) %>% 
    dplyr::arrange(padj) %>%
        head(10)
```

## Visualize DE effect


### Visualize top two hits


Looking at the results for these two genes: 

* The estimated log2FC for "ENSMUSG00000071724.6" is positive. We will verify visually that treatment is associated with higher expression
* The estimated log2FC for "ENSMUSG00000035683.14" is negative. We will verify visually that treatment is associated with lower expression


```{r}
DESeq2::plotCounts(ddsobj, "ENSMUSG00000071724.6", intgroup = "trt", normalized = TRUE)

DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>%
    dplyr::filter(row == "ENSMUSG00000071724.6")
```

```{r}
DESeq2::plotCounts(ddsobj, "ENSMUSG00000035683.14", intgroup = "trt", normalized = TRUE)

DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>%
    dplyr::filter(row == "ENSMUSG00000035683.14")
```

```{r}
myinteractplot <- function(mydds, geneid) {
    SummarizedExperiment::assay(mydds) %>%
        tibble::as_tibble(rownames="gene") %>%
            dplyr::filter(gene == geneid) %>%
                tidyr::gather(Label, value, -gene) %>%
                    dplyr::select(-gene) -> 
                        genedat

    print(genedat)
    SummarizedExperiment::colData(mydds) %>%
        as.data.frame %>%
            tibble::as_tibble() %>%
                dplyr::full_join(genedat, by = "Label") -> genedat
    
    genedat %>%
        ggplot2::ggplot(aes(x = trt, y = value, color = cline)) +
            ggplot2::scale_colour_manual(name = "",  values = c("red3", "blue3")) +
                ggplot2::theme_bw()
    }


#myinteractplot(ddsobj, "ENSMUSG00000071724.6")
```


### Volcano plot

```{r}
### Volcano plot for treatment effect
DESeq2::results(ddsDE, contrast = c("trt", "antiPD1", "isotype"), tidy = TRUE) %>%
    dplyr::filter(!is.na(pvalue)) %>%
        ggplot2::ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) + 
            geom_point()
```

```{r}
### Volcano plot for cell line effect
DESeq2::results(ddsDE, contrast = c("cline", "UV2", "Parental"), tidy = TRUE) %>%
    dplyr::filter(!is.na(pvalue)) %>%
        ggplot2::ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) + 
            geom_point()
```

```{r}
sessionInfo()
```
