---
title: DEseq2 differential expression analysis
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.11.2
  kernelspec:
    display_name: R
    language: R
    name: ir
---


## Objective: Carry out a basic set of DE analyses using DESeq2 and visualize the results


### Load packages

```{r}
library(tidyverse)
library(DESeq2)
```

### Load the PRJNA668393 dds object from image file

```{r}
procdir <- normalizePath("./proc/")
ddsfile <- file.path(procdir, "PRJNA668393-ddsdata.RDS")

PRJNA668393ddsData <- readRDS(ddsfile)

tools::md5sum(ddsfile)
```

# Differential expression analysis


## Specify Model


You recall that DESeq requires that  we have estimates for sample specific size factors and gene specific dispersion factors. More specifically, recall that DESeq models the count $K_{ij}$ (gene $i$, sample $j$) as negative binomial with mean $\mu_{ij}$ and dispersion parameter $\alpha_i$. Here $\mu_{ij}=s_j q_{ij}$ where $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} x_j + \beta_{2i} z_j + \beta_{3i} (x_j*z_j)$. Here $s_j$ is the sample $j$ specific size factor.

**Summarize of notation**
- $K_{ij}$ denotes the observed **number of reads** mapped to gene $i$ for sample $j$
- $K_{ij}$ follows a **negative binomial distribution** with
    - **Mean** $\mu_{ij}$
    - **Dispersion parameter** $\alpha_i$
- Modelling
    - $K_{ij} \sim NB(\mu_{ij}, \alpha_i)$
    - $\mu_{ij} = s_{j}q_{ij}$
        - $s_j$ is sample $j$ specific normalization constant
    - $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} x_j + \beta_{2i} z_j $
        - $x_j=0$ if treament = isotype, or
        - $x_j=1$ if treatment = anti-PD1
        - $z_j=0$ if cell line = Parental, or
        - $z_j=1$ if cell line = UV2.
 - Effect size (no cell line effect)
     - If there is differential treatment effect but NO cell line effect for gene $i$, then
         - $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i}$
     - The corresponding mean count is
         - $\mu_{ij} = s_j 2^{\beta_{0i} + \beta_{1i}}$
 
  - Effect size (with cell line effect)
     - If there is differential treatment and cell line effects for gene $i$, then
         - $\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} + \beta_{2i}$
     - The corresponding mean count is
         - $\mu_{ij} = s_j 2^{\beta_{0i} + \beta_{1i} \beta_{2i}}$   


Add the design to the DESeq2 object

```{r}
ddsobj <- PRJNA668393ddsData
```

```{r}
design(ddsobj) <- formula(~ cline + trt + cline:trt)
```

```{r}
design(ddsobj)
```

## 01 Size Factors
 We begin by estimating the size factors $s_1,\ldots,s_n$:

```{r}
ddsobj <- estimateSizeFactors(ddsobj)
```

Now, compare the dds object to that of before applying the estimateSizeFactors() function. What has changed? What remains unchanged?

```{r}
ddsobj
```

Note that there is a **sizeFactor** added to **colData**. Let's look at it more carefully

You can also get the size factors directly

```{r}
tibble::tibble(sfactor = sizeFactors(ddsobj))
```

Summarize size factors

```{r}
summary(sizeFactors(ddsobj))
```

Now that the size factors have been estimated, we can get "normalized" counts (DESeq2 normalizes against size factor)


## 02 Dispersion Parameters
Next, we get the dispersion factors $\alpha_1,\ldots,\alpha_{m}$

```{r}
ddsobj <- estimateDispersions(ddsobj)
```

Now inspect the dds object again and note that the rowRanges slot has extra information ("metadata column names(0):" before versus "column names(9): baseMean baseVar ... dispOutlier dispMAP")
- before: 
    - `metadata column names(0):`
- after:  
    - `column names(9): baseMean baseVar ...`

```{r}
alphas <- dispersions(ddsobj)
```

Verify that the number of dispersion factors equals the number of genes

```{r}
summary(alphas)
```

Summarize the dispersion factors using a box plot (may want to log transform)

```{r}
boxplot(log(dispersions(ddsobj)))
```

# Differential Expression Analysis
We can now conduct a differential expression analysis using the DESeq() function. Keep in mind that to get to this step, we first estimated the size factors and then the dispersion parameters.

```{r}
### Carry out interaction analysis
ddsDE <- DESeq2::DESeq(ddsobj)
```

```{r}
### Look at object
ddsDE
```

```{r}

DESeq2::resultsNames(ddsDE)
```

```{r}
### Look at some of the results
DESeq2::results(ddsDE)
```

### Look at some of the results (tidy version sorted by adjusted P-value)

```{r}
DESeq2::results(ddsDE, tidy = TRUE) %>% 
    dplyr::arrange(padj) %>%
        head(10)
```

We can get the results for the differential expression analysis using results(). Here, we can compare two group of samples specified by the contrast. (If not, the default contrast would be the last term in your additive model `design(dds)`).


## Extract coefficient estimates


The results for two genes are shown below

```{r}
DESeq2::results(ddsDE, tidy = TRUE) %>%
    dplyr::filter(row %in% c("ENSMUSG00000022351.15","ENSMUSG00000055653.14"))
```

This output only provides the estimate for the treament effect $\beta_{1i}$ for this gene. We can get the estimates for the intercept $\beta_{0i}$ and the cell line effect $\beta_{2i}$ as follows

```{r}
coef(ddsDE)[c("ENSMUSG00000022351.15","ENSMUSG00000055653.14") , ]
```

## Visualize DE effect


### Visualize top two hits

```{r}
myinteractplot <- function(mydds, geneid) {
    
    SummarizedExperiment::assay(mydds) %>%
        tibble::as_tibble(rownames="gene") %>%
            dplyr::filter(gene == geneid) %>%
                tidyr::gather(Run, value, -gene) %>%
                    dplyr::select(-gene) -> 
                        expdat
    
    SummarizedExperiment::colData(mydds) %>%
        as.data.frame %>%
            tibble::as_tibble() %>%
                dplyr::full_join(expdat, by = "Run") -> 
                    genedat
    
    genedat %>%
        ggplot2::ggplot(aes(x = cline, y = value/sizeFactor, color = trt)) +
            ggplot2::geom_point() + 
                ggplot2::xlab("Cell line") + ggplot2::ylab(paste(geneid, "(normalized count)")) +
                ggplot2::scale_colour_manual(name = "", values = c("red3", "blue3")) +
                    ggplot2::theme_bw()
    }

myinteractplot(ddsobj, "ENSMUSG00000022351.15")

myinteractplot(ddsobj, "ENSMUSG00000055653.14")
```


Bonus exercise: The previous helper function allowed us to customize the differential expression visualization. It is however in need of improvement and refinement if to be used for other projects. Specifically, there is substantial hard coding (e.g., the variable name for the merger "Run", the experimental factors "trt" and "cline", the axis labels, the colors. Consider revising this function so that it can be used more generally.


### Volcano plot

```{r}
### Volcano plot for treatment effect
DESeq2::results(ddsDE, tidy = TRUE) %>%
    dplyr::filter(!is.na(pvalue)) %>%
        ggplot2::ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) + 
            ggplot2::geom_point() + 
                ggplot2::theme_bw()
```

```{r}
sessionInfo()
```
